torchrl/modules/planners/mpc.py

```diff
@@ -0,0 +1,186 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+import abc
+from typing import Optional
+
+import torch
+from torchrl.data.tensordict.tensordict import TensorDictBase
+from torchrl.envs import EnvBase
+from torchrl.modules import TensorDictModule
+
+
+class MPCPlannerBase(TensorDictModule, metaclass=abc.ABCMeta):
+    """
+    MPCPlannerBase Module. This class inherits from TensorDictModule. This is an abstract class and must be implemented by the user.
+    Provided a TensorDict, this module will perform a Model Predictive Control (MPC) planning step.
+    At the end of the planning step, the MPCPlanner will return the action that should be taken.
+
+    Args:
+        env (Environment): The environment to perform the planning step on (Can be ModelBasedEnv or EnvBase).
+        action_key (str): The key in the TensorDict to use to store the action.
+
+    Returns:
+        TensorDict: The TensorDict with the action added.
+    """
+
+    def __init__(
+        self,
+        env: EnvBase,
+        action_key: str = "action",
+    ):
+        out_keys = [action_key]
+        in_keys = [env.input_spec.keys()]
+        super().__init__(env, in_keys=in_keys, out_keys=out_keys)
+        self.env = env
+        self.action_spec = env.action_spec
+
+    def _call_module(self, tensordict: TensorDictBase) -> torch.Tensor:
+        return self.planning(tensordict)
+
+    @abc.abstractmethod
+    def planning(self, tensordict: TensorDictBase) -> torch.Tensor:
+        """
+        Perform the MPC planning step.
+        Args:
+            tensordict (TensorDict): The TensorDict to perform the planning step on.
+        Returns:
+            TensorDict: The TensorDict with the action added.
+        """
+        raise NotImplementedError()
+
+    def forward(
```

Comment:
why can't we use the plain forward here?
There's little interest in inheriting from TensorDictModule if we're rewriting fundamental methods.
I thought re-writing _call_module only was sufficient?
(if it's for the functionality side of things, I guess we can just raise errors if vmap or params are passed?)

Reply:
The problem is that the original __call_modules takes as input tensors and ours takes tensordicts. Do we don't want to perform ```tensors = tuple(tensordict.get(in_key, None) for in_key in self.in_keys)```

Reply:
Ok, perhaps we don't need the _call_module then? Why do we need this?
```python
    def _call_module(self, tensordict: TensorDictBase) -> torch.Tensor:
        return self.planning(tensordict)
```